org/mozilla/jss/ssl/SSLSocket.c

```diff
@@ -630,6 +656,14 @@ Java_org_mozilla_jss_ssl_SSLSocket_socketConnect
         goto finish;
     }
 
+    /* Set ALPN extension */
+    if (alpn != NULL) {
+        JSS_RefByteArray(env, alpn, (jbyte**)&alpnData, &alpnLength);
+        if (alpnLength > 0) {
+            SSL_SetNextProtoNego(sock->fd, alpnData, alpnLength);
```

Comment:
ALPN data only really needs to be provided before the handshake is performed. Not that I mind the extra constructors, but should we add a separate `setApplicationProtocols(...)` function to add this after construction? This would let us set it after construction from a `JSSSocketFactory` (which I think we have in ldap-jdk and/or PKI).

Reply:
(And IIRC, it can be used with TLSv1.2 and can be provided on later handshakes. Under TLSv1.3 there's only the initial one so I guess it doesn't matter too much.)

Reply:
I don't think we need to worry about the renegotiation use case.  But if it does turn out to be important we could add the extra function then.

Or now... I do prefer supplying as much "config" in the constructor rather than afterwards.  Having to call extra functions to do more setup is something I find distasteful (I use that word in acknowledgement that it is indeed a matter of taste ^_^).  Ideally there would be a single configuration record type that can supply SNI, ALPN and be extensible as new knobs are needed, so that there is  not a combinatorial explosion in the number of possible constructors :)

Reply:
It also turns out (I had forgotten until I made changes in my other PR) -- NSS doesn't support changing ALPN values on renegotiation. Even 0RTT assumes (and validates!) ALPN doesn't change any.

> "single configuration record type"
... So ... [JSSParameters](https://dogtagpki.github.io/jss/master/javadocs/org/mozilla/jss/ssl/javax/JSSParameters.html)? :-)

Reply:
@cipherboy yep :D